iree/compiler/Dialect/VM/Conversion/VMToEmitC/ConvertVMToEmitC.cpp

```diff
@@ -94,60 +209,247 @@ Optional<emitc::ApplyOp> createVmTypeDefPtr(ConversionPatternRewriter &rewriter,
+  }
+
+  rewriter.setInsertionPointToEnd(condBlock);
+  auto branchOp = rewriter.create<CondBranchOp>(
+      location, conditionI1.getResult(0), continuationBlock, failureBlock);
+
+  rewriter.setInsertionPoint(continuationBlock, opPosition);
+
   return callOp;
 }
 
-// TODO(simon-camp/marbre): Use this function throughout the conversions.
-Optional<std::string> getCType(Type type) {
-  if (auto iType = type.dyn_cast<IntegerType>()) {
-    switch (iType.getWidth()) {
-      case 32:
-      case 64:
-        return std::string("int") + std::to_string(iType.getWidth()) +
-               std::string("_t");
+emitc::CallOp returnIfError(ConversionPatternRewriter &rewriter,
+                            Location location, StringAttr callee,
+                            ArrayAttr args, ArrayAttr templateArgs,
+                            ArrayRef<Value> operands) {
+  auto blockBuilder = [&rewriter, &location](emitc::CallOp &callOp) {
+    auto ctx = rewriter.getContext();
+
+    Block *block = rewriter.getBlock();
+    mlir::FuncOp funcOp = cast<mlir::FuncOp>(block->getParentOp());
+
+    // Release local refs
+    for (auto constantOp : funcOp.getOps<emitc::ConstantOp>()) {
+      Operation *op = constantOp.getOperation();
+      if (!op->hasAttr("ref_ordinal")) continue;
+
+      auto refPtrOp = rewriter.create<emitc::ApplyOp>(
+          /*location=*/location,
+          /*type=*/emitc::OpaqueType::get(ctx, "iree_vm_ref_t*"),
+          /*applicableOperator=*/StringAttr::get(ctx, "&"),
+          /*operand=*/constantOp.getResult());
+
+      rewriter.create<emitc::CallOp>(
+          /*location=*/location,
+          /*type=*/TypeRange{},
+          /*callee=*/StringAttr::get(ctx, "iree_vm_ref_release"),
+          /*args=*/ArrayAttr{},
+          /*templateArgs=*/ArrayAttr{},
+          /*operands=*/ArrayRef<Value>{refPtrOp.getResult()});
     }
-  }
 
-  if (auto fType = type.dyn_cast<FloatType>()) {
-    switch (fType.getWidth()) {
-      case 32:
-        return std::string("float");
-      case 64:
-        return std::string("double");
+    rewriter.create<mlir::ReturnOp>(location, callOp.getResult(0));
+  };
+
+  auto ctx = rewriter.getContext();
+  Type type = emitc::OpaqueType::get(ctx, "iree_status_t");
+  return failableCall(rewriter, location, type, callee, args, templateArgs,
+                      operands, blockBuilder, /*negateResult=*/true);
+}
+
+emitc::CallOp failListNull(ConversionPatternRewriter &rewriter,
+                           Location location, Type type, StringAttr callee,
+                           ArrayAttr args, ArrayAttr templateArgs,
+                           ArrayRef<Value> operands) {
+  auto blockBuilder = [&rewriter, &location](emitc::CallOp &callOp) {
+    auto ctx = rewriter.getContext();
+
+    Block *block = rewriter.getBlock();
+    mlir::FuncOp funcOp = cast<mlir::FuncOp>(block->getParentOp());
+
+    // Release local refs
+    for (auto constantOp : funcOp.getOps<emitc::ConstantOp>()) {
```

Comment:
since this is done in ~6 places it'd be good to abstract it out to some releaseLocalRefs helper

Reply:
Yes, this needs a ton of cleanup, I just wanted to get this out before I head into the weekend :D
So a look on the overall structure would be enough for now :)

Reply:
it looks great! very cool to see the fidelity improvement such that some of the runtime helpers can be removed